File: scripts/oop-practice-sources/lua/shop-cart.lua
local Product = {}

function Product:new(name, price)
    local obj = setmetatable({name = name, price = price}, {__index = self})
    return obj
end

local Cart = {}

function Cart:new()
    local obj = setmetatable({items = {}}, {__index = self})
    return obj
end

function Cart:add(product)
    table.insert(self.items, product)
    print("В корзину добавлено: " .. product.name .. " (" .. product.price .. " ₽)")
end

function Cart:total()
    local sum = 0
    for _, p in ipairs(self.items) do
        sum = sum + p.price
    end
    return sum
end

local Order = {}

function Order:new(cart)
    local items = {}
    for _, item in ipairs(cart.items) do
        table.insert(items, item)
    end
    local obj = setmetatable({items = items, total = cart:total()}, {__index = self})
    return obj
end

function Order:checkout()
    print("Оформление заказа...")
    for _, item in ipairs(self.items) do
        print("  — " .. item.name .. ": " .. item.price .. " ₽")
    end
    print("Итого: " .. self.total .. " ₽")
    print("Заказ оформлен!")
end

local cart = Cart:new()
cart:add(Product:new("Книга", 500))
cart:add(Product:new("Ручка", 50))
local order = Order:new(cart)
order:checkout()

File: scripts/oop-practice-sources/java/shop-cart.java
import java.util.ArrayList;

class Product {
    String name;
    int price;

    Product(String name, int price) {
        this.name = name;
        this.price = price;
    }
}

class Cart {
    ArrayList<Product> items = new ArrayList<>();

    void add(Product product) {
        items.add(product);
        System.out.println("В корзину добавлено: " + product.name + " (" + product.price + " ₽)");
    }

    int total() {
        int sum = 0;
        for (Product p : items) sum += p.price;
        return sum;
    }
}

class Order {
    ArrayList<Product> items;
    int total;

    Order(Cart cart) {
        this.items = new ArrayList<>(cart.items);
        this.total = cart.total();
    }

    void checkout() {
        System.out.println("Оформление заказа...");
        for (Product item : items) {
            System.out.println("  — " + item.name + ": " + item.price + " ₽");
        }
        System.out.println("Итого: " + total + " ₽");
        System.out.println("Заказ оформлен!");
    }
}

public class Main {
    public static void main(String[] args) {
        Cart cart = new Cart();
        cart.add(new Product("Книга", 500));
        cart.add(new Product("Ручка", 50));
        Order order = new Order(cart);
        order.checkout();
    }
}
